### ISSUE TITLE
Help Wanted: Unit & Functional Tests

### ISSUE THREAD

**Issue by bigtimebuddy** [pixijs/pixi.js#3047]
Hello Pixi Devs!

We're looking for some PRs from folks to help increase our test-coverage or Pixi.js. Why are unit-tests helpful? Many reasons, but here are a few: 
- they help us spot issues sooner (with Travis and running `npm test`)
- they improve the stability of code over time
- they aid in doing major refactoring efforts

Let's be frank, our test-coverage right now is _pretty_ poor. Until now, we hadn't realized just how poor. We have just merged PR #3041 a code-coverage tool which analyzes the Pixi code against the unit-tests to see what was run and what wasn't (uses [istanbul](http://gotwarlost.github.io/istanbul/)). 

Right now our current **coverage is about 32%**. You can view the **[latest report](https://s3-eu-west-1.amazonaws.com/pixi.js/dev/coverage/index.html)** on the current **dev** branch. We really would like to see that number grow over time. So.... we will try to improve our coverage by at least 10% for the next couple releases. 

Basically, we need some help writing some tests. Here are some of the lower-hanging fruit that's currently not being unit-tested. Here are some ideas, would gladly welcome PRs to unit-test these pieces:
### Math
- [x] [Matrix](https://pixi-builds.s3-eu-west-1.amazonaws.com/dev/coverage/src/core/math/Matrix.js.html)
- [ ] [GroupD8](https://pixi-builds.s3-eu-west-1.amazonaws.com/dev/coverage/src/core/math/GroupD8.js.html)
- [x] [Point](https://pixi-builds.s3-eu-west-1.amazonaws.com/dev/coverage/src/core/math/Point.js.html)
- [x] [ObservablePoint](https://pixi-builds.s3-eu-west-1.amazonaws.com/dev/coverage/src/core/math/ObservablePoint.js.html)
### Math Shapes
- [x] [Circle](https://pixi-builds.s3-eu-west-1.amazonaws.com/dev/coverage/src/core/math/shapes/Circle.js.html)
- [x] [Ellipse](https://pixi-builds.s3-eu-west-1.amazonaws.com/dev/coverage/src/core/math/shapes/Ellipse.js.html)
- [ ] [Polygon](https://pixi-builds.s3-eu-west-1.amazonaws.com/dev/coverage/src/core/math/shapes/Polygon.js.html)
- [x] [Rectangle](https://pixi-builds.s3-eu-west-1.amazonaws.com/dev/coverage/src/core/math/shapes/Rectangle.js.html)
- [ ] [RoundedRectangle](https://pixi-builds.s3-eu-west-1.amazonaws.com/dev/coverage/src/core/math/shapes/RoundedRectangle.js.html)
### Loader Middleware
- [x] [bitmapFontParser](https://pixi-builds.s3-eu-west-1.amazonaws.com/dev/coverage/src/loaders/bitmapFontParser.js.html)
- [x] [spritesheetParser](https://pixi-builds.s3-eu-west-1.amazonaws.com/dev/coverage/src/loaders/spritesheetParser.js.html)
- [x] [textureParser](https://pixi-builds.s3-eu-west-1.amazonaws.com/dev/coverage/src/loaders/textureParser.js.html)

**Comment by ivanpopelyshev** [comment]
Test the point. Ugh.

I can cover textureParser and GroupD8, those are my things :)

**Comment by bigtimebuddy** [comment]
Thanks Ivan :+1:

**Comment by staff0rd** [comment]
What do the red/yellow/green highlights mean within the modules/classes?

**Comment by GoodBoyDigital** [comment]
Nice one @bigtimebuddy ! I agree we should definitely aim to increase this!

**Comment by bigtimebuddy** [comment]
@staff0rd Green is good means the line was run, the number on the left is the number of times it was run. 

Yellow means a branch was not hit. For instance, if only once case of a switch/case is hit, the others would show up yellow. So technically the code is run, but not all the conditions are run. 

Red means not run.

**Comment by staff0rd** [comment]
Is there any intention of moving the tests to ES6?

**Comment by englercj** [comment]
They were ES5 previously because they ran in a few browsers. Then i kept them ES5 thinking we would eventually do something like SauceLabs or BrowserStack and since they would run in our lowest common denominator they would need to be ES5.

The reality today is that currently they run in electron, and if we did switch to SauceLabs or such in the future, we could transpile them before running. I think @bigtimebuddy already changed the linter to allow ES6 in tests, so go for it.

**Comment by staff0rd** [comment]
When I run the coverage command locally, it only produces the base index.html, how can I make it do the rest?

**Comment by bigtimebuddy** [comment]
What do you mean @staff0rd? Are there any other files in the coverage directory?

**Comment by staff0rd** [comment]
I run `npm run coverage` and these appear in `/coverage/`;
~~~
base.css
coverage-final.json
index.html
prettify.css
prettify.js
sort-arrow-sprite.png
sorter.js
~~~

When I browse the `index.html` all the links throw 404s, because they haven't been generated.  For example, `/source/src/accessibility/index.html`.

**Comment by bigtimebuddy** [comment]
That's so weird. Are you on Windows?

**Comment by staff0rd** [comment]
Yah Windows 10

**Comment by staff0rd** [comment]
Also, I added tests for Container.setChildIndex but it still displays in yellow [here](https://s3-eu-west-1.amazonaws.com/pixi.js/dev/coverage/src/core/display/Container.js.html).  Am I doing something wrong?

**Comment by godric3** [comment]
I'm writing some tests for math shapes and just to be sure:
`PIXI.Rectangle(0, 0, 1, 1)` is expected to NOT contain points (0,1), (1,0) and (1,1)?

**Comment by englercj** [comment]
Looks like the algorithm currently does not count points *on* the border as "contained". So yes, that is current behavior.

**Comment by ivanpopelyshev** [comment]
may be make it inclusive?

**Comment by geordidearns** [comment]
Hey all, I'm trying to get started writing some tests for the Ellipse math object, but can't seem to run any tests on my forked repo. I'm getting this error, any pointers? Thanks

`[floss] caught outer exception: Cannot find module '../dist/pixi'`

**Comment by staff0rd** [comment]
@geordidearns run the build first;
~~~
npm run build
~~~

**Comment by geordidearns** [comment]
thanks @staff0rd, got it going 👍

**Comment by out-of-band** [comment]
I stumbled on your Issues page and thought I'd take a crack at filling out your test suite a bit.  Starting simple, and aiming to get some of the basic shapes up to 100% line coverage and improve the branch coverage where possible.

I have two questions (hopefully this is the right place to ask) regarding the Rectangle class:

- What is the intended functionality of the `Rectangle.fit(rectangle)` method?  Given the following:

```
    const rect1 = new PIXI.Rectangle(0, 0, 5, 5);
    const rect2 = new PIXI.Rectangle(10, 10, 10, 10);

    rect1.fit(rect2);
```

... what _should_ be the left/top/right/bottom values for rect1?  This might also be a good place to beef up documentation.

- If the tests uncover redundant code or code that could be refactored, would you prefer to those changes to be made on a separate PR?

Thanks!

**Comment by bigtimebuddy** [comment]
**GREAT NEWS** was have really improved our test coverage since starting this thread it has increased nearly 10%. I'm going to close this post for now and start another thread with some new unit-testing needs. Thanks for everyone's help. Please keep adding to our tests and submitting PRs.
